### PR TITLE
Enable blob caching and blobInfo caching

### PR DESCRIPTION
**Pull request by nalind**
This patch set updates the versions of the storage, image, and buildah libraries and enables blob and blobInfo caching.  We switch back to the main containers/image repository, since it now supports registry names specified as IP addresses, and drop logic for setting the InsecureSkipTLSVerify flag, which the image library's registry client logic now sets automatically if we don't.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -24,3 +25,6 @@ RUN ln -s /usr/bin/openshift-builder /usr/bin/openshift-sti-build && \
 LABEL io.k8s.display-name="OpenShift Builder" \
       io.k8s.description="This is a component of OpenShift and is responsible for executing image builds." \
       io.openshift.tags="openshift,builder"
+
+RUN mkdir /var/cache/blobs
+VOLUME /var/lib/containers/cache
```
Comment (adambkaplan) [inline Dockerfile]: Is the intent that `/var/lib/containers/cache` will be mounted in by the build controller (from the host)?

File: pkg/build/builder/cmd/builder.go [hunk 2]
```diff
@@ -131,7 +132,18 @@ func newBuilderConfigFromEnvironment(out io.Writer, needsDocker bool) (*builderC
 		}
 		istorage.Transport.SetStore(store)
 
-		dockerClient, err := bld.GetDaemonlessClient(systemContext, store, os.Getenv("BUILD_ISOLATION"))
+		// By default, enable a blob cache, but allow it to be moved or
+		// disabled via the environment.
+		cfg.blobCache = "/var/cache/blobs"
+		if dir, isSet := os.LookupEnv("BUILD_BLOBCACHE"); isSet {
```
Comment (adambkaplan) [inline pkg/build/builder/cmd/builder.go]: How will this be disabled - empty env valule?

File: pkg/build/builder/daemonless.go [hunk 3]
```diff
@@ -68,28 +62,17 @@ func pullDaemonlessImage(sc types.SystemContext, store storage.Store, imageName
...
-			glog.V(2).Infof("Registry %q is marked as insecure in the registries configuration.", registry.URL)
-			systemContext.DockerInsecureSkipTLSVerify = true
-			systemContext.OCIInsecureSkipTLSVerify = true
-		} else {
-			glog.V(2).Infof("Registry %q is marked as secure in the registries configuration.", registry.URL)
-		}
-	} else {
-		glog.V(2).Infof("Registry for %q is not present in the registries configuration, assuming it is secure.", imageName)
-	}
-
```
Comment (adambkaplan) [inline pkg/build/builder/daemonless.go]: I assume the designation of secure/insecure registries will be taken care of by copying in `/etc/containers/registries.conf`?

See https://github.com/openshift/origin/pull/21653

File: Dockerfile (hunk 1 above)
Comment (bparees) [reply Dockerfile]: it is, but i don't think we gain anything by declaring this volume, i'd rather leave it out.  We mount stuff in lots of places in the image that aren't explicitly declared as volumes.

@nalind was there a specific reason you needed to add the volume declaration?

File: pkg/build/builder/cmd/builder.go (hunk 2 above)
Comment (bparees) [reply pkg/build/builder/cmd/builder.go]: please call this BUILD_ENABLE_BLOBCACHE or something that makes it clear it's an enablement (and then make it enable when the value is set to `true` (case insensitive))

File: pkg/build/builder/cmd/builder.go [hunk 4]
```diff
@@ -131,7 +132,18 @@ func newBuilderConfigFromEnvironment(out io.Writer, needsDocker bool) (*builderC
...
+			cfg.blobCache = dir
+		}
+
+		// An empty value implies the system default location, somewhere under
+		// /var/lib/containers/cache.
+		systemContext.BlobInfoCacheDir = os.Getenv("BUILD_BLOBINFOCACHE")
```
Comment (bparees) [inline pkg/build/builder/cmd/builder.go]: and name this one BUILD_BLOBINFOCACHE_DIR or something that makes it clear it points to a path.

Comment (bparees) [reply pkg/build/builder/cmd/builder.go]: also presumably you need to handle the case that someone enabled the blob cache but did not set this dir.

or do away w/ this env entirely and just hardcode the expected cache directory.

File: pkg/build/builder/daemonless.go (hunk 3 above)
Comment (bparees) [reply pkg/build/builder/daemonless.go]: @nalind did something get fixed in buildah that we don't need this any more?

File: Dockerfile (hunk 1 above)
Comment (nalind) [reply Dockerfile]: It's a hint that there's a benefit to mounting `/var/lib/containers/cache` as a shared volume, but you're right that we don't do that for the other things that we mount in, so I'll drop the line.

File: pkg/build/builder/cmd/builder.go (hunk 2 above)
Comment (nalind) [reply pkg/build/builder/cmd/builder.go]: Okay, changing it.

File: pkg/build/builder/daemonless.go (hunk 3 above)
Comment (nalind) [reply pkg/build/builder/daemonless.go]: Yes, https://github.com/containers/image/pull/468 was merged.

Comment (nalind) [reply pkg/build/builder/daemonless.go]: @adambkaplan Yup, the way the builder expects the configuration to be provided to it hasn't changed, we just don't need to do this bit of work ourselves any more.

File: pkg/build/builder/cmd/builder.go (hunk 2 above)
Comment (bparees) [reply pkg/build/builder/cmd/builder.go]: sorry I screwed this up.  but i think BUILD_BLOBCACHE_DIR would be a better name.

Comment (bparees) [reply pkg/build/builder/cmd/builder.go]: (and ignore my comments about expecting a value of `true`)

File: pkg/build/builder/cmd/builder.go (hunk 4 above)
Comment (nalind) [reply pkg/build/builder/cmd/builder.go]: Okay, dropping the variable, which should result in containers/image always using its default location.

**Comment by nalind** [comment]
Backing out the changes to pay attention to the ImageOptimizationPolicy, which can be done in a different PR after this one.

**Comment by runcom** [comment]
This still needs a corresponding origin PR for the controller (e.g. https://github.com/openshift/origin/pull/21613)

**Comment by runcom** [comment]
picked up this PR to bring in https://github.com/containers/buildah/pull/1244 for testing till it gets merged and I'm seeing errors coming from pkg/errors and raven-go. I've been able to overcome that by sticking the lib in glide with:
```
+- package: github.com/pkg/errors
+  version: 645ef00459ed84a119197bfb8d8205042c6df63d
+- package: github.com/getsentry/raven-go
+  version: 32a13797442ccb601b11761d74232773c1402d14
```
it is related to https://github.com/getsentry/raven-go/issues/225

**Comment by runcom** [comment]
The parallel pull + pgzip PR in Buildah has been merged. We could grab it here, I'm already testing with it.

**Comment by nalind** [comment]
> We need [containers/buildah#1265](https://github.com/containers/buildah/pull/1265) as well

Included as of https://github.com/openshift/builder/pull/34#event-2060437413.

File: Dockerfile [hunk 5]
```diff
@@ -16,6 +16,8 @@ COPY --from=builder /go/src/github.com/openshift/builder/openshift-builder /usr/
 COPY imagecontent/policy.json /etc/containers/
 COPY imagecontent/registries.conf /etc/containers/
 COPY imagecontent/storage.conf /etc/containers/
+RUN mkdir /var/cache/blobs
```
Comment (bparees) [inline Dockerfile]: There's some discussion about changing this name to "var/cache/blobs/metadata" assuming this is the metadata cache.  Please sync with @runcom and @mrunalp

**Comment by bparees** [comment]
one nit on the dir name, otherwise lgtm pending passing tests.

File: Dockerfile (hunk 5 above)
Comment (runcom) [reply Dockerfile]: related to https://github.com/openshift/origin/pull/21613#discussion_r245744542

Comment (runcom) [reply Dockerfile]: `/var/cache/blobs/meta` I've used that in 21613 suggested by @mrunalp

File: pkg/build/builder/cmd/builder.go [hunk 6]
```diff
@@ -131,7 +132,15 @@ func newBuilderConfigFromEnvironment(out io.Writer, needsDocker bool) (*builderC
 		}
 		istorage.Transport.SetStore(store)
 
-		dockerClient, err := bld.GetDaemonlessClient(systemContext, store, os.Getenv("BUILD_ISOLATION"))
+		// Default to using /var/cache/blobs as a blob cache, but allow its location
+		// to be changed by setting $BUILD_BLOBCACHE_DIR.  Setting the location to an
+		// empty value disables the cache.
+		cfg.blobCache = "/var/cache/blobs"
```
Comment (bparees) [inline pkg/build/builder/cmd/builder.go]: this default should also change of course

File: pkg/build/builder/cmd/builder.go [hunk 7]
```diff
@@ -131,7 +132,15 @@ func newBuilderConfigFromEnvironment(out io.Writer, needsDocker bool) (*builderC
...
+		if blobCacheDir, isSet := os.LookupEnv("BUILD_BLOBCACHE_DIR"); isSet {
```
Comment (bparees) [inline pkg/build/builder/cmd/builder.go]: and this env var name should change to indicate it's metadata

File: Dockerfile (hunk 5 above)
Comment (bparees) [reply Dockerfile]: nm, this dir is apparently for something else.

**Pull request merged**